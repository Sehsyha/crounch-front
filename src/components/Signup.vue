<template>
  <div class="spinner">
    <img src="../assets/loading.svg" alt="loading"/>
  </div>
</template>

<script>
import router from './../router';

export default {
  name: 'callback',
  props: ['auth'],
  data() {
    this.axios.post('http://localhost:3000/users').then(() => {
      router.replace({path: '/'})
    })
    return {}
  }
}
</script>

<style scoped>
 .spinner {
    position: absolute;
    display: flex;
    justify-content: center;
    height: 100vh;
    width: 100vw;
    background-color: white;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
  }
</style>
