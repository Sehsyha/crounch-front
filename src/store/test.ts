import { ActionTree, MutationTree, GetterTree, Module, ActionContext } from 'vuex';
import { RootState } from '.';

export type ActionCall<S, R> = (injectee: ActionContext<S, R>, payload?: any) => any;

export function createStoreModuleMock<S, A, M, G>(initialState: S, actionKeys: A, mutationKeys: M, getterKeys: G): Module<S, RootState> {
  const actions: ActionTree<S, RootState> = {};
  const mutations: MutationTree<S> = {};
  const getters: GetterTree<S, RootState> = {};

  Object.keys(actionKeys).forEach(key => actions[key] = jest.fn());
  Object.keys(mutationKeys).forEach(key => mutations[key] = jest.fn());
  Object.keys(getterKeys).forEach(key => getters[key] = jest.fn());

  return {
    namespaced: true,
    actions,
    getters,
    mutations,
    state: initialState,
  }
}
