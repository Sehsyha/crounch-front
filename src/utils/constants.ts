export const TOKEN_STORAGE_KEY = 'token';

export const DEFAULT_CATEGORY_ID = '00000000-0000-0000-0000-000000000000'
export const DEFAULT_CATEGORY_NAME = 'Divers'
