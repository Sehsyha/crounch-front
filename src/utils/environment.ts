export function getAPIURL(): string {
  return process.env.NODE_ENV === 'production' ? 'http://crounch.me:3000/' : 'http://localhost:3000/';
}
