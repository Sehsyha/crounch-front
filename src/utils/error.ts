export interface ErrorBody {
  code: string;
  description: string;
}

export class FetchError extends Error {
  status: number;
  code: string;
  description: string;

  constructor(message: string, status: number, body: ErrorBody) {
    super(message)
    this.status = status
    this.code = body.code
    this.description = body.description
  }
}
