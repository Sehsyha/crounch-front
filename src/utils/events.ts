export enum Events {
  ADD_PRODUCT = 'add-product',
  DELETE_PRODUCT = 'delete-product'
}
