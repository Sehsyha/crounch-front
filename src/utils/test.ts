import Vue from 'vue';
import { shallowMount, VueClass } from '@vue/test-utils';

export function shallowComponent<V extends Vue>(component: VueClass<V>, values: {} = {}) {
  return shallowMount(component, {
    attachToDocument: true,
    propsData: {
      ...values
    }
  });
}
