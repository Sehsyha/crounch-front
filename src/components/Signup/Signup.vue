<template>
  <div id="signup">
    <h2>Signup</h2>
    <form @submit.prevent="signup">
      <div v-if="!isEmailValid" class="error" id="email-error">Please enter a valid email.</div>
      <div v-if="!isPasswordValid" class="error" id="password-error">Please enter a password.</div>
      <div v-if="signupSuccess" class="success">You've signed up successfully !!</div>
      <input type="text" placeholder="Email" v-model="email" />
      <input type="password" placeholder="Password" v-model="password" />
      <input type="submit" value="S'inscrire" />
    </form>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import { Action } from 'vuex-class';

import { validateEmail } from '@/utils/form-validation';
import { AuthKeys } from '@/store/auth/keys';
import { authNamespace } from '@/store/auth';

@Component
export default class Signup extends Vue {
  @Action(AuthKeys.SIGNUP, authNamespace) doSignup: any;

  email: string = '';
  password: string = '';
  signupSuccess: boolean = false;

  signup() {
    if (!this.isEmailValid || !this.isPasswordValid) {
      return;
    }

    const { email, password } = this;
    this.doSignup({ email, password }).then(() => {
      this.signupSuccess = true;
    });
  }

  get isEmailValid(): boolean {
    return !!validateEmail(this.email);
  }

  get isPasswordValid(): boolean {
    return this.password.length !== 0;
  }
}
</script>
