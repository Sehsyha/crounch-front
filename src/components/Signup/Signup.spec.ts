import { Wrapper } from '@vue/test-utils';
import { when } from 'jest-when';
import axios from 'axios';

import Signup from '@/components/Signup/Signup.vue';
import { shallowComponent } from '../../utils/test';
import { validateEmail } from '../../utils/form-validation';

jest.mock('axios', () => ({
  post: jest.fn(),
}));

jest.mock('../../utils/form-validation')

describe('Signup', () => {
  let wrapper: Wrapper<Signup>;
  const email = 'a';
  const password = 'password';

  beforeEach(() => {
    (axios.post as jest.Mock).mockClear();
    (axios.post as jest.Mock).mockReturnValue(Promise.resolve({}));
    wrapper = shallowComponent(Signup);
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('Should render.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('Init', () => {
    it('Should initialize email to empty string.', () => {
      expect(wrapper.vm.$data.email).toBe('');
    });

    it('Should initialize password to empty string.', () => {
      expect(wrapper.vm.$data.password).toBe('');
    });

    it('Should initialize signupSuccess to false.', () => {
      expect(wrapper.vm.$data.signupSuccess).toBe(false);
    });
  });

  describe('Form validation', () => {
    describe('Email', () => {
      it('Should render an error when email is not valid.', () => {
        when(validateEmail as jest.Mock).calledWith(email).mockReturnValue(false);
        wrapper.setData({ email });
        expect(wrapper.find('#email-error').exists()).toBeTruthy();
      });

      it('Should not render an error when email is valid.', () => {
        when(validateEmail as jest.Mock).calledWith(email).mockReturnValue(true);
        wrapper.setData({ email });
        expect(wrapper.find('#email-error').exists()).toBeFalsy();
      });
    });

    describe('Password', () => {
      it('Should render an error when password input is not filled.', () => {
        expect(wrapper.find('#password-error').exists()).toBeTruthy();
      });

      it('Should not render an error when password input is filled.', () => {
        wrapper.setData({ password: 'az' });

        expect(wrapper.find('#password-error').exists()).toBeFalsy();
      });
    });
  });

  describe('Form submission', () => {
    it('Should not call api when email is in error and not password.', () => {
      when(validateEmail as jest.Mock).calledWith(email).mockReturnValue(false);
      wrapper.setData({ password: 'password', email });

      wrapper.find('input[type=submit]').trigger('click');

      expect(axios.post as jest.Mock).not.toHaveBeenCalled();
    });

    it('Should not call api when password is in error and not email.', () => {
      when(validateEmail as jest.Mock).calledWith(email).mockReturnValue(true);
      wrapper.setData({ password: '', email });

      wrapper.find('input[type=submit]').trigger('click');

      expect(axios.post as jest.Mock).not.toHaveBeenCalled();
    });

    it('Should call api with right parameters when everything is valid.', () => {
      when(validateEmail as jest.Mock).calledWith(email).mockReturnValue(true);
      wrapper.setData({ password, email });

      wrapper.find('[type=submit]').trigger('click');

      expect(axios.post as jest.Mock).toHaveBeenCalledWith('http://localhost:3000/users', { email, password });
    });

    it('Should display success message when user has signed up.', done => {
      when(validateEmail as jest.Mock).calledWith(email).mockReturnValue(true);
      wrapper.setData({ password, email });

      wrapper.find('[type=submit]').trigger('click');

      setTimeout(() => {
        expect(wrapper.find('.success').exists()).toBeTruthy();
        done();
      });
    });
  });
});
