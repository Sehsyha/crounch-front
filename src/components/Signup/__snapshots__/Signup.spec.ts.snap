// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Should render. 1`] = `
<form>
  <div id="email-error" class="error">Please enter a valid email.</div>
  <div id="password-error" class="error">Please enter a password.</div>
  <!----> <input type="text" placeholder="Email"> <input type="password" placeholder="Password"> <input type="submit" value="S'inscrire">
</form>
`;
